core/factorization/symbolic.cpp

```diff
@@ -67,38 +67,41 @@ GKO_REGISTER_HOST_OPERATION(compute_elim_forest, compute_elim_forest);
 /** Computes the symbolic Cholesky factorization of the given matrix. */
 template <typename ValueType, typename IndexType>
 void symbolic_cholesky(
-    const matrix::Csr<ValueType, IndexType>* mtx,
+    xstd::dont_deduce_t<pointer_param<const matrix::Csr<ValueType, IndexType>>>
```

Comment:
this is a specific property of C++ template argument deduction. If you don't want to know, feel free to stop reading.

Template arguments are deduced (with a few exceptions, that is references, const-ness, array-to-pointer, function-to-pointer, derived-to-base-class-pointer conversions etc.) to be exact matches of what we pass in. So when we pass in std::unique_ptr, it is not an exact match to gko::pointer_param (they are different classes), as no implicit conversions will be allowed. So we need to make sure that this parameter doesn't participate in deduction. How do we do that? In come class templates!
Class templates can be specialized in C++, which is what makes a lot of the SFINAE tooling like std::enable_if work. The fact that they can be specialized means that type aliases that occur inside class templates can't be resolved unless the class template parameter is fixed. Since it is to be deduced here, the whole parameter will just be ignored in type deduction, and ValueType and IndexType will be deduced from the second parameter.

Reply:
Thanks a lot for the explanation! This is a very interesting usage that I did not know about templates.